kernel/core/datamanagement/memory.cpp

```diff
@@ -3748,94 +3748,162 @@ std::vector<mu::value_type> Memory::getIndex(size_t col, const std::vector<mu::v
+
+            //todo this is done to often, how more efficient
+            groupedValues[1].memArray.push_back(TblColPtr(_mem.memArray[0]->copy(VectorIndex(&catIndex2[0], catIndex2.size(), 0))));
+            // todo is there any chance to sort mu::value_type ?
+            //std::sort(catIndex1.begin(), catIndex1.end());
+            //std::sort(catIndex2.begin(), catIndex2.end());
+            std::vector<mu::value_type> intersection;
+
+            //std::set_intersection(catIndex1.begin(), catIndex1.end(), catIndex2.begin(), catIndex2.end(),
+            //                      std::back_inserter(intersection));
+
+            for(auto a : catIndex1)
+                for(auto b : catIndex2)
+                    if(a == b)
+                        intersection.push_back(a);
+
+            groupedValues[2].memArray.push_back(TblColPtr(_mem.memArray[0]->copy(VectorIndex(&intersection[0], intersection.size(), 0))));
+        }
     }
 
     // Prepare vectors for each group
-    std::vector<mu::value_type> vAvg;
-    std::vector<mu::value_type> vVar;
-    std::vector<mu::value_type> vNum;
-
-    // Get the values for each group
-    for (size_t j = 2; j < _mem.memArray.size(); j++)
+    std::vector<std::vector<mu::value_type>> vAvg_(groupedValues.size());
+    std::vector<std::vector<mu::value_type>> vVar_(groupedValues.size());
+    std::vector<std::vector<mu::value_type>> vNum_(groupedValues.size());
+    for(size_t i = 0; i < groupedValues.size(); i++)
     {
-        vAvg.push_back(_mem.avg(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(j)));
-        vNum.push_back(_mem.num(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(j)));
-        vVar.push_back(intPower(_mem.std(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(j)), 2));
+        for (size_t j = 0; j < groupedValues[i].memArray.size(); j++)
+        {
+            vAvg_[i].push_back(groupedValues[i].avg(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(j)));
+            vNum_[i].push_back(groupedValues[i].num(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(j)));
+            vVar_[i].push_back(intPower(groupedValues[i].std(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(j)), 2));
+        }
     }
 
     // Calculate the overall values
-    mu::value_type overallAvg = _mem.avg(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(2, VectorIndex::OPEN_END));
-    mu::value_type overallNum = _mem.num(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(2, VectorIndex::OPEN_END));
-    mu::value_type overallVariance;
+    mu::value_type overallAvg = _mem.avg(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(0));
+    mu::value_type overallNum = _mem.num(VectorIndex(0, VectorIndex::OPEN_END), VectorIndex(0));
 
-    // Pretend that each group contains n equal measurements
-    // to calculate the ideal overall variance
-    for (size_t i = 0; i < vAvg.size(); i++)
+    //calculate SS_Within
+    mu::value_type SS_Within;
+    for(size_t i = 0; i < vAvg_.back().size(); i++)

```

Comment:
I've some doubts about using `back()` here, which refers to the interaction component. Using `back()` here instead of using some named object hides a awesome bug, if we want to extend that to more dimensions.

Reply:
hmmm I thought that this is already the right approach for any dimension. 

> SSW is one component of total sum of squares (the other is between sum of squares). Within sum of squares represents the variation due to individual differences in the score. In other words, it’s the variation of individual scores around the group mean;

and the last element in the vector should correspond to the grouping using all Categories.
Or did I understand something wrong here?

Reply:
I meant, although it is more expressive than `vector[vector.length()-1]`, `vector.back()` does not carry a literal meaning for the algorithm itself, i.e., you have to *know and ensure* that the last element in the vector is the interaction. `vector.back()` does not tell you that.

Reply:
Understand, good point! How would you suggest implementing it?

Reply:
Good question. To be as portable as possible, what about having a recursive structure? E.g. something as follows:

```c++
struct RecursiveData
{
    RecursiveData* m_interaction = nullptr;
    double m_val1 = 0.0;
    double m_val2 = 0.0;

    ~RecursiveData()
    {
        if (m_interaction)
            delete m_interaction;
    }
    
    double getInteraction()
    {
        if (m_interaction)
            return m_interaction->getInteraction();
        
        return m_val1*m_val2; // only a dummy calculation
    }
};